### PR TITLE
Refactor filesystem package

### PR DESCRIPTION
**Pull request by ktock**
We refactor it to the following components for better maintainability,
- reader:  A rich wrapper for CRFS's stargz.Reader. We can use it for opening
           files in stargz archive with cache, prefetching them, etc. Previously
           this functionality was implemented in "filereader.go".
- remote:  A NW-aware package for reading a blob on remote URL. Previously, this
           functionality was implemented in "urlreader.go".
- handler: [unchanged] A handler package for containerd client.

### CODE REVIEW

File: stargz/fs.go [hunk 1]
```diff
@@ -103,141 +110,156 @@ func getCache(ctype, dir string, maxEntry int) (cache.BlobCache, error) {
...
 	if err != nil {
 		return nil, err
 	}
+	interval := config.LayerValidInterval
+	if interval == 0 {
+		// zero means "use default interval"
+		interval = int64(defaultLayerValidInterval)
+	} else if interval < 0 {
+		// negative value means "check every time"
+		interval = 0
```
Comment (AkihiroSuda) [inline stargz/fs.go]: This logic smells. I suggest adding a new config field.

File: stargz/fs.go [hunk 2]
```diff
@@ -281,74 +321,40 @@ func (fs *filesystem) Mount(ctx context.Context, mountpoint string, labels map[s
...
-// e.g. using HTTP instead of HTTPS.
-func (fs *filesystem) isInsecure(host string) bool {
-	for _, i := range fs.insecure {
-		if ok, _ := regexp.Match(i, []byte(host)); ok {
-			return true
-		}
-	}
-
-	return false
+func (fs *filesystem) unmount(mountpoint string) {
```
Comment (AkihiroSuda) [inline stargz/fs.go]: this function name is confusing, it doesn't actually unmount?

File: stargz/reader/reader.go [hunk 3]
```diff
@@ -79,42 +68,51 @@ func (gr *stargzReader) openFile(name string) (io.ReaderAt, error) {
...
 		}
 		return nil
 	}
 
 	return
 }
 
-func (gr *stargzReader) cacheTarGz(r io.Reader) error {
+func (gr *Reader) FetchTarGzWithReader(sr *io.SectionReader) error {
+	r := bufio.NewReaderSize(sr, 2<<28)
```
Comment (AkihiroSuda) [inline stargz/reader/reader.go]: 1<<29 ?

**Comment by ktock** [comment]
Thanks for the review. I fixed and will marge it.

**Pull request merged**